Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM_2nd.cpp

```diff
@@ -0,0 +1,540 @@
+            M_iter_maxerror = -1.0;
+
+            // loop over cells and update fields
+            amrex::ParallelFor(tbx, tby, tbz,
+            [=] AMREX_GPU_DEVICE (int i, int j, int k){
+
+              // when working on M_xface(i,j,k, 0:2) we have direct access to M_xface(i,j,k,0:2) and Hx(i,j,k)
+              // Hy and Hz can be acquired by interpolation
+              // H_maxwell
+              Real Hx_xface = MacroscopicProperties::getH_Maxwell(i, j, k, 0, amrex::IntVect(1,0,0), amrex::IntVect(1,0,0), Bx, M_prev_xface);
+              Real Hy_xface = MacroscopicProperties::getH_Maxwell(i, j, k, 1, amrex::IntVect(0,1,0), amrex::IntVect(1,0,0), By, M_prev_xface);
+              Real Hz_xface = MacroscopicProperties::getH_Maxwell(i, j, k, 2, amrex::IntVect(0,0,1), amrex::IntVect(1,0,0), Bz, M_prev_xface);
+              // H_bias
+              Real Hx_bias_xface = MacroscopicProperties::face_avg_to_face(i, j, k, 0, amrex::IntVect(1,0,0), amrex::IntVect(1,0,0), Hx_bias);
+              Real Hy_bias_xface = MacroscopicProperties::face_avg_to_face(i, j, k, 0, amrex::IntVect(0,1,0), amrex::IntVect(1,0,0), Hy_bias);
+              Real Hz_bias_xface = MacroscopicProperties::face_avg_to_face(i, j, k, 0, amrex::IntVect(0,0,1), amrex::IntVect(1,0,0), Hz_bias);
+              // H_eff = H_maxwell + H_bias + H_exchange + H_anisotropy ... (only the first two terms are considered here)
+              Real Hx_eff = Hx_bias_xface;
+              Real Hy_eff = Hy_bias_xface;
+              Real Hz_eff = Hz_bias_xface;
+
+              // Real Hx_eff = Hx_xface + Hx_bias_xface;
+              // Real Hy_eff = Hy_xface + Hy_bias_xface;
+              // Real Hz_eff = Hz_xface + Hz_bias_xface;
+
+              // magnetic material properties mag_alpha and mag_Ms are defined at cell nodes
+              // keep the interpolation
+              Real mag_gamma_interp = MacroscopicProperties::macro_avg_to_face(i,j,k,amrex::IntVect(1,0,0),mag_gamma_arr);
+
+              Real a_temp_dynamic_coeff = PhysConst::mu0 * std::abs(mag_gamma_interp) *
+                        (1.0 + std::pow(MacroscopicProperties::macro_avg_to_face(i,j,k,amrex::IntVect(1,0,0),mag_alpha_arr), 2.0))/ 2.0;
+
+          // calculate a_temp_xface
+              // x component on x-faces of grid
+              a_temp_xface(i, j, k, 0) = -( dt * a_temp_dynamic_coeff * Hx_eff + a_temp_static_xface(i, j, k, 0) );
+
+              // y component on x-faces of grid
+              a_temp_xface(i, j, k, 1) = -( dt * a_temp_dynamic_coeff * Hy_eff + a_temp_static_xface(i, j, k, 1) );
+
+              // z component on x-faces of grid
+              a_temp_xface(i, j, k, 2) = -( dt * a_temp_dynamic_coeff * Hz_eff + a_temp_static_xface(i, j, k, 2) );
+
+              // calculate M_xface
+              // x component on x-faces of grid
+              M_xface(i, j, k, 0) = MacroscopicProperties::getM_half (i, j, k, 0, a_temp_xface, b_temp_xface);
+
```

Comment:
from the way you calculate the coefficient for `a`, I thought you want to update `M_integer` at this step, so that in WarpXEvolve.cpp, you are using `MacroscopicEvolveM_2nd(0.5*dt[0])`? Why `M_half`?

Reply:
Maybe we should have a comment saying the intended usage is to pass in 0.5*dt into this routine, so the notion of Mhalf implies half dt step knowing this?  I don't feel to strongly either way as long as there is a comment about what is going on.

Reply:
I changed this function to updateM_field and add some comments to make it clear.